connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceTask.java

```diff
@@ -69,14 +69,19 @@ public MirrorSourceTask() {}
 
     // for testing
     MirrorSourceTask(KafkaConsumer<byte[], byte[]> consumer, MirrorSourceMetrics metrics, String sourceClusterAlias,
-                     ReplicationPolicy replicationPolicy, long maxOffsetLag, KafkaProducer<byte[], byte[]> producer) {
+                     ReplicationPolicy replicationPolicy, long maxOffsetLag, KafkaProducer<byte[], byte[]> producer,
+                     Semaphore outstandingOffsetSyncs, Map<TopicPartition, PartitionState> partitionStates,
```

Comment:
I'm not sure we should be mocking the semaphore here, since it's a bit of an implementation detail and I've already proposed that we split it out into several different semaphores (one for each topic partition) in https://github.com/apache/kafka/pull/12366.

Can we get the same testing coverage/guarantees by mocking out the offset sync producer instead, and selectively invoking/not invoking the callbacks passed to it in `send`?

Reply:
I have updated the MR without mocking the outstanding syncs semaphore but I have preserved the semaphore constructor parameter. The semaphore is getting initialized in the `start` method and I could not find a better way to avoid the NPE.

Reply:
One alternative is to hardcode the instantiation of the `outstandingOffsetSyncs` semaphore in the testing-only constructor, without relying on a constructor parameter (and the same can be done for the `partitionStates` field).

But that has its own problems of duplicating instantiation logic for those fields and leading to possible divergence in behavior between task instances depending on whether they're brought up for testing or not, if we're not careful about keeping that instantiation logic in sync.

One way that that issue can be addressed is by pulling out any non-trivial instantiation logic into a separate static method; e.g.:
```java
private static Semaphore newOutstandingOffsetSyncsSemaphore() {
    return new Semaphore(MAX_OUSTANDING_OFFSET_SYNCS);
}
```

But I don't think that this is so much better than what you have right now that I'd block the PR on adapting this approach; even if it has some advantages, the additional complexity makes it debatable whether it'd really be worth the tradeoff.